### ISSUE TITLE
Implemented configurable API endpoints for chat models

### ISSUE THREAD

**Issue by liumengdi** [brianpetro/obsidian-smart-connections#104]
can we add support for customize API address url?  so we can replace https://api.openai.com with custom proxy server

**Comment by brianpetro** [comment]
Hi @liumengdi 

This is currently not on the roadmap. I have some questions.

In what cases might someone need to do this? In the case that it's to avoid geographic restrictions, is it ethical to do this? Why or why not?

Are there other ways that sufficiently determined users can do this? Perhaps using a VPN?

Thanks,
Brian🌴

**Comment by abhuva** [comment]
> 

i could imagine that this way we could use other services like HugginFace or a self-run service like LLama.

**Comment by brianpetro** [comment]
@abhuva unfortunately it isn't that simple because there isn't any sort of standardized method for access the models :/

**Comment by riverai** [comment]
> Hi @liumengdi
> 
> This is currently not on the roadmap. I have some questions.
> 
> In what cases might someone need to do this? In the case that it's to avoid geographic restrictions, is it ethical to do this? Why or why not?
> 
> Are there other ways that sufficiently determined users can do this? Perhaps using a VPN?
> 
> Thanks, Brian🌴

Hello.
Microsoft Cloud Platform is now available to request api for openai gpt3.5. Currently, there is already a program that can forward data sent to OpenAI to the Microsoft api, provided that the program can customize the api address.

Moreover, this api is still free to use for one year.
At the same time, Microsoft's gpt api service is open to more countries around the world.
If this has any ethical issues, obviously this is something Microsoft needs to think about.


So I hope that the ability to customize the api address will go on the roadmap. Thanks!

**Comment by runchard** [comment]
Hi @brianpetro 

> Are there other ways that sufficiently determined users can do this? Perhaps using a VPN?

For one, is that Azure has provided an same API structure but just with different base URL. Azure's version is designed for enterprise users with better privacy support.

Furthermore, through a simple proxy layer can assist individuals in effectively managing and monitoring API data at the application level. (At the moment, I have many services relying on one OpenAI account. However, OpenAI only provides a monthly usage limit. Once the limit is exceeded, all other services will be affected.)

VPN is not enough to do so.

It is nice to have this feature.

**Comment by murazi** [comment]
+1

It is suggested to set a separate API parsing address for "openai" because of the access mechanism. I am using a third-party "openai" platform, so the API parsing address needs to be specially handled.

**Comment by pvginkel** [comment]
I'm also looking for this. I can't use this plugin for my work notes because my company doesn't allow me to share company information with OpenAI.

It seems that the API's are mostly the same, except that obviously the URL is different, and that an API version is expected. See https://github.com/openai/openai-python/blob/fe3abd16b582ae784d8a73fd249bcdfebd5752c9/openai/api_resources/abstract/createable_api_resource.py#L29 for an example.

It'd be terrific if you could add support for this. I'm already considering automatically creating MD files of all my incoming email so that I can search that also. It would help me so much with my daily tasks.

**Comment by atmomomo** [comment]
Support. Some users (like me) use GPT via Azure instead. Thanks!

**Comment by brianpetro** [comment]
@atmomomo can you confirm whether the azure API endpoints are the same as OpenAI? Mainly, are the request/response parameters the same for both the embedding and chat completion APIs? And does the azure API use a simple API key for auth?

Thanks
Brian 🌴

**Comment by pvginkel** [comment]
The endpoints are largely the same. They're described here:

https://learn.microsoft.com/en-us/azure/cognitive-services/openai/reference

The differences as far as I understand them:

- API token is passed in through a header named `api-key`.
- Model is **not** passed in as part of the request bodies. Instead, Azure has the concept of an implementation that you pass in as part of the URL. In the end, it's quite immaterial. What I've done in my compatibility layer is that if I detect an Azure setup, I clear the model from the request body and put it into the query string. The full name of the model must then be user configurable, because it won't be the exact model name. I have e.g. named my 3.5 turbo `default-gpt-35-turbo`.
- The URL's are built up as follows:
  - Resource URL
  - `openai/deployments/`
  - Model name (they call this a `deployment-id`)
  - The rest is the same as the OpenAI URL without the version
- And then every request needs to have a `version` query string parameter. The current version is `2023-05-15`.

Long story short, lets say the OpenAI request is:

```
POST https://api.openai.com/v1/embeddings

Authorization: Bearer 12345

{
  "model": "gpt-3.5-turbo",
  ...
}
```

The Azure URL would be:

```
POST https://my-openai-resource.openai.azure.com/openai/deployments/default-gpt-35-turbo/completions?api-version=2023-05-15

api-key: 12345

{
  // No model property
  ...
}
```

Assuming the name of the resource is `my-openai-resource` and the name of the model is `default-gpt-35-turbo`. I would make the full URL origin configurable, not just the resource part, so `https://my-openai-resource.openai.azure.com/`.

The only real difference I found in the implementation is that the Azure API only supports one embedding per request. The rest seems to work as is.

**Comment by Mingzefei** [comment]
Hello @brianpetro ,
Is there any update on whether this feature request is still being considered for future development? I really hope to have this functionality as it would greatly enhance the flexibility of the chat models.

**Comment by brianpetro** [comment]
@Mingzefei I was just working on this a few minutes ago before checking these notifications 😊

**Comment by brianpetro** [comment]
Custom API settings available since `v2.1` 🌴

**Comment by brianpetro** [comment]
Recorded response https://youtu.be/OoYChPfdR5U
